common/src/typed_uuid.rs

```diff
@@ -0,0 +1,26 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use newtype_uuid::{TypedUuidKind, TypedUuidTag};
+use schemars::JsonSchema;
+
+macro_rules! impl_typed_uuid_kind {
+    ($($kind:ident => $tag:literal),* $(,)?) => {
+        $(
+            #[derive(JsonSchema)]
+            pub enum $kind {}
+
+            impl TypedUuidKind for $kind {
+                #[inline]
+                fn tag() -> TypedUuidTag {
+                    TypedUuidTag::new($tag)
+                }
+            }
```

Comment:
I don't fully understand this. Why do we need a unique enum for every type of uuid? (I suspect this is just a rust pattern I'm ignorant of, I'd love to learn more)

Reply:
The idea behind the PR is to have a marker type which indicates that a UUID represents a particular identity. The marker type can be anything that implements `TypedUuidKind`. So we'd have a type for sled IDs, one for project IDs, etc, etc. Once we have that, we can rely on the compiler to check that we aren't mixing up different kinds of IDs.

Now, note that the marker type does not need to store any data, so a zero-sized type is ideal. One option is an empty struct, e.g. `MyKind(())`, and that's fine. But in our case we don't actually need to ever construct an instance of that type! So, in the interest of making illegal states unrepresentable, we're looking to define a type which _cannot be constructed_.

What kind of type can never be constructed? The [never type (`!`)](https://doc.rust-lang.org/std/primitive.never.html) is one answer, certainly, but in this case we want to define lots of types that are different but cannot be constructed. The answer is that we define an enum without any variants. (It is easy to see that the enum can never be constructed -- if there aren't any variants, how would you specify one?)

This ties into some ideas behind type systems. You might have heard that a struct is called a "product type" and an enum is called a "sum type". One way to see how these names came about is to look at the cardinality of these types--i.e. how many values each type can take.

For example:

```rust
struct MyStruct(u8, bool);

enum MyEnum {
    Integer(u8),
    Boolean(bool),
}
```

How many different values can `MyStruct` and `MyEnum` represent?

* For `MyStruct`, there's two elements, a `u8` (256 elements) and a `bool` (2 elements). The total number of different values `MyStruct` can take is **256 * 2 = 512**.
* For `MyEnum`, the value can either be a `u8` or a `bool`. The total number of different values is **256 + 2 = 258**.

Now consider an "empty" struct and an "empty" enum.

```rust
struct EmptyStruct;

enum EmptyEnum {}
```

So, what's the cardinality of `EmptyStruct`? It's the identity for multiplication, which is **1**. `EmptyStruct` can only take one possible value; it can only be defined one way.

And similarly, the cardinality of `EmptyEnum`, being the identity for addition, is **0**. In other words, `EmptyEnum` can simply never take a value. It must always be zero.

---

There is a downside to this, though, which is that we can't define a UUID that's "either this or that", e.g. "either a sled or a project". For that, we'd need to move beyond marker types -- we'd need to store some kind of value along with the type. Practical use cases for this aren't entirely clear to me, but if you (or someone else) has some we should switch over to that. (It would complicate our APIs a bit, so if we don't need that I'd rather we not have that.)

(Apologies if this was too long-winded and you were aware of most or all of it!)

Reply:
Very instructive, thank you!

Reply:
I wrote up https://github.com/oxidecomputer/newtype-uuid/issues/7 as a response to the "why marker types" question.